src/wallet/wallet.cpp

```diff
@@ -2403,34 +2403,68 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
     return res;
 }
 
-bool CWallet::SignTransaction(CMutableTransaction& tx)
+bool CWallet::SignTransaction(CMutableTransaction& tx) const
 {
     AssertLockHeld(cs_wallet);
 
-    // sign the new tx
-    int nIn = 0;
+    // Build coins map
+    std::map<COutPoint, Coin> coins;
     for (auto& input : tx.vin) {
         std::map<uint256, CWalletTx>::const_iterator mi = mapWallet.find(input.prevout.hash);
         if(mi == mapWallet.end() || input.prevout.n >= mi->second.tx->vout.size()) {
             return false;
         }
-        const CScript& scriptPubKey = mi->second.tx->vout[input.prevout.n].scriptPubKey;
-        const CAmount& amount = mi->second.tx->vout[input.prevout.n].nValue;
-        SignatureData sigdata;
+        const CWalletTx& wtx = mi->second;
+        coins[input.prevout] = Coin(wtx.tx->vout[input.prevout.n], wtx.m_confirm.block_height, wtx.IsCoinBase());
+    }
+    std::map<int, std::string> input_errors;
+    return SignTransaction(tx, coins, SIGHASH_ALL, input_errors);
+}
 
-        std::unique_ptr<SigningProvider> provider = GetSigningProvider(scriptPubKey);
-        if (!provider) {
-            // We don't know about this scriptpbuKey;
-            return false;
+bool CWallet::SignTransaction(CMutableTransaction& tx, const std::map<COutPoint, Coin>& coins, int sighash, std::map<int, std::string>& input_errors) const
+{
+    // Sign the tx with ScriptPubKeyMans
+    // Because each ScriptPubKeyMan can sign more than one input, we need to keep track of each ScriptPubKeyMan that has signed this transaction.
+    // Each iteration, we may sign more txins than the txin that is specified in that iteration.
+    std::set<uint256> visited_spk_mans;
+    for (unsigned int i = 0; i < tx.vin.size(); i++) {
+        // Get the prevout
+        CTxIn& txin = tx.vin[i];
+        auto coin = coins.find(txin.prevout);
+        if (coin == coins.end() || coin->second.IsSpent()) {
+            input_errors[i] = "Input not found or already spent";
+            continue;
         }
 
-        if (!ProduceSignature(*provider, MutableTransactionSignatureCreator(&tx, nIn, amount, SIGHASH_ALL), scriptPubKey, sigdata)) {
-            return false;
+        // Check if this input is complete
+        SignatureData sigdata = DataFromTransaction(tx, i, coin->second.out);
+        if (sigdata.complete) {
+            continue;
         }
-        UpdateInput(input, sigdata);
-        nIn++;
+
+        // Input needs to be signed, find the right ScriptPubKeyMan
+        ScriptPubKeyMan* spk_man = GetScriptPubKeyMan(coin->second.out.scriptPubKey, sigdata);
```

Comment:
Will one SPKM always be sufficient for any input even going forward? Seems to me the same SPKM will always be sourced for a particular input?

Reply:
5cab890b915fd9f25cada7ddab7cc00eea22c80b suggests no, there may be multiple.

Reply:
The expectation is that one spkman will be sufficient. But I suppose it is possible that multiple spkmans could sign for a particular input. I don't think we handle that case.

5cab890 was from a previous iteration where all spkmans signed the transaction. Perhaps it isn't needed anymore.

Reply:
I think it makes sense to remove that commit then, and just state the assumption here.

Reply:
Dropped the commit. Added a comment about the assumption.

Reply:
I've had to restore the commit. It actually is needed. It actually doesn't imply that the different `ScriptPubKeyMan`s can successfully sign the same input.

It is still needed because each time the wrong `ScriptPubKeyMan` signs an input, it will add an error. Once that input becomes signed by the correct `ScriptPubKeyMan`, we need to clear away those errors otherwise they will incorrectly persist. Thus that commit is still needed.

Reply:
Shouldn't this loop over all matching `GetScriptPubKeyMan`s? Otherwise the comment above should be changed to "We assume that each input is **matched** (and signed) by only one ScriptPubKeyMan.".

Right now:
* if non-matching `ScriptPubKeyMan` tries and fails to sign this input, we'll try with the matching `ScriptPubKeyMan`, but;
* if we match, but can't sign, that error can only be cleared if another input happens to match a `ScriptPubKeyMan` that then happens to be able for the failed input as well.

Reply:
Fine. I've changed it to check all `ScriptPubKeyMan`s that `CanProvide`.